consensus/tendermint/committee/committee_set_test.go

```diff
@@ -0,0 +1,393 @@
+		assertNilError(t, err)
+
+		setSize := set.Size()
+		if int64(setSize) != n {
+			t.Fatalf("expected committee set size: %v and got: %v", n, setSize)
+		}
+	}
+
+	for _, size := range committeeSetSizes {
+		size := size
+		t.Run(fmt.Sprintf("committee size of %v member/s", size), func(t *testing.T) {
+			assertSetSize(t, size)
+		})
+	}
+
+}
+
+func TestSet_Committee(t *testing.T) {
+	var committeeSetSizes = []int64{1, 2, 10, 100}
+	var assertSetCommittee = func(t *testing.T, n int64) {
+		committeeMembers := createTestCommitteeMembers(t, n, genRandUint64(int(n), maxSize))
+		set, err := NewSet(copyMembers(committeeMembers), committeeMembers[0].Address)
+		sort.Sort(committeeMembers)
+		assertNilError(t, err)
+
+		gotCommittee := set.Committee()
+
+		if !reflect.DeepEqual(committeeMembers, gotCommittee) {
+			t.Fatalf("expected committee: %v and got: %v", committeeMembers, gotCommittee)
+		}
+	}
+
+	for _, size := range committeeSetSizes {
+		size := size
+		t.Run(fmt.Sprintf("get committee of %v member/s", size), func(t *testing.T) {
+			assertSetCommittee(t, size)
+		})
+	}
+}
+
+func TestSet_GetByIndex(t *testing.T) {
+	committeeMembers := createTestCommitteeMembers(t, 4, genRandUint64(4, maxSize))
+	sort.Sort(committeeMembers)
+	set, err := NewSet(copyMembers(committeeMembers), committeeMembers[0].Address)
+	assertNilError(t, err)
+
+	t.Run("can get member by index", func(t *testing.T) {
+		expectedMember := committeeMembers[1]
+		gotMember, err := set.GetByIndex(1)
+		assertNilError(t, err)
+
+		if !reflect.DeepEqual(expectedMember, gotMember) {
+			t.Fatalf("expected member: %v and got %v", expectedMember, gotMember)
+		}
+	})
+
+	t.Run("error on accessing member index not in committee", func(t *testing.T) {
+		_, err := set.GetByIndex(6)
+		assertError(t, consensus.ErrCommitteeMemberNotFound, err)
+	})
+}
+
+func TestSet_GetByAddress(t *testing.T) {
+	committeeMembers := createTestCommitteeMembers(t, 4, genRandUint64(4, maxSize))
+	sort.Sort(committeeMembers)
+	set, err := NewSet(copyMembers(committeeMembers), committeeMembers[0].Address)
+	assertNilError(t, err)
+
+	t.Run("can get member by Address", func(t *testing.T) {
+		expectedMember := committeeMembers[1]
+		index, gotMember, err := set.GetByAddress(expectedMember.Address)
+		assertNilError(t, err)
+
+		if index != 1 {
+			t.Fatalf("incorrect index of member expected: %v and got %v", 1, index)
+		}
+
+		if !reflect.DeepEqual(expectedMember, gotMember) {
+			t.Fatalf("expected member: %v and got %v", expectedMember, gotMember)
+		}
+	})
+
+	t.Run("error on accessing member address not in committee", func(t *testing.T) {
+		_, _, err := set.GetByAddress(common.HexToAddress("testaddress"))
+		assertError(t, consensus.ErrCommitteeMemberNotFound, err)
+	})
+}
+
+// TestSet_GetProposer tests the round robin selection of proposers by continuously calling GetProposer and checking
+// the resulting Committee is the same as ordered Committee passed to the set
```

Comment:
I was thinking we could make reference to the properties that are being checked, something like:
```suggestion
// TestSet_GetProposer tests the round robin selection of proposers. It validates that as GetProposer is
// called with consecutive rounds, consecutive proposers are chosen in the sort order defined by
// types.Committee. The consequence of this is that proposers are selected fairly, with N-1 other proposers
// being selected between any two instances of the same proposer being selected twice. It also validates that
// the selection process starts from the committee member that follows lastBlockProposer in a sorted instance
// of types.Committee.
```

The last sentence of this comment I think is something that is currently missing from this test, but should be added.

Reply:
The last sentence is not necessary if you think it is please checkout the branch and make a commit

Reply:
Just to clarify my thinking, if we had the following members `ABCDEF` this test could pass in its current form if every round (of round robin) the sequence `ABCDEF` was returned by consecutive calls to `GetProposer` irrespective of what `lastBlockProposer` was provided at construction time.

If this is correct then I think we need that extra check.